libcontainer/rootfs_linux.go

```diff
@@ -73,6 +73,8 @@ func prepareRootfs(pipe io.ReadWriter, iConfig *initConfig, mountFds []int) (err
 		// Therefore, we can access mountFds[i] without any concerns.
 		if mountFds != nil && mountFds[i] != -1 {
 			mountConfig.fd = &mountFds[i]
+		} else {
+			mountConfig.fd = nil
```

Comment:
I don't follow how this fixes the issue. Can you please elaborate?

mountConfig.fd is a `*int` (see [here][mountConfigFd]) and mountConfig is created in the scope of this function [here][mountConfig]. Then, the `else` should be a no-op, as pointers are nil initialized in go.

I also checked and prepareRottfs() is called from one place only, so... I really don't see how this can help.

I'm sure I'm missing something, but what is it? 

[mountConfigFd]: https://github.com/opencontainers/runc/pull/3510/files#diff-e0932af06cb46065b2ef5fc89bc2bd9b880c9578e01b89ef4ecaeffb78a8bef0R37
[mountConfig]: https://github.com/opencontainers/runc/pull/3510/files#diff-e0932af06cb46065b2ef5fc89bc2bd9b880c9578e01b89ef4ecaeffb78a8bef0L63-L68

Reply:
@rata The variable `mountConfig` is defined outside the "for" loop, so `mountConfig.fd` might carry over the value set at a previous iteration of the loop.

Reply:
Right! So, as we talked, it seems moving the mountConfig inside the loop will fix that and similar bugs in the future too.

Will open a PR with that approach